### PR TITLE
add more common types

### PR DESCRIPTION
**Pull request by jgravois**
porting a whole load of interfaces and types originally written by @JeffJacobson :tada: 

AFFECTS PACKAGES:
@esri/arcgis-rest-common-types
@esri/arcgis-rest-geocoder
@esri/arcgis-rest-groups
@esri/arcgis-rest-items
batch-geocoder

BREAKING CHANGE:
no longer prefacing interface names with an `I`

maybe its just the ArcObjects phobia talking, but 5 minutes into this port i started to agree with [@tomwayson](https://github.com/Esri/arcgis-rest-js/issues/107#issuecomment-367217416). A quick look at the TypeScript Language [Specification](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#7.1) made it clear that its not a universally agreed upon convention.

i'm planning on tackling #104 soon so it makes sense to me to rip off the proverbial esri-loader band-aid here too and just release a `2.0.0` if everyone else agrees.

for now, the associated linter rule is turned [`off`](https://palantir.github.io/tslint/rules/interface-name/) because we're still using the `I` outside common-types.

### CODE REVIEW

**Comment by patrickarlt** [comment]
@jgravois @tomwayson while prefixing interfaces with `I` isn't a universal standard by any means (it is borrowed from .Net) but think it makes sense to differentiate between classes in the JS API `Extent`, `Point`, `SpatialReference` and `Item` and the interfaces that we use in these libraries, `IExtent`, `IPoint`, `ISpatialReference` and `IItem`.

I would hate to end up with something like this:


```ts
import { Extent } from "@esri/rest-js-common-types";

require("esri/Extent", (Extent) => {
  // now you have 2 problems...
});
```

Prefixing with `I` quickly tells you that the thing you are looking at is an interface and not a class. Just because the [TypeScript Guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines) discourage it is no reason not to do since the guidelines are for _contributing to TypeScript_ not writing TypeScript code in general.

File: packages/arcgis-rest-groups/src/groups.ts [hunk 1]
```diff
@@ -230,9 +231,9 @@ export function updateGroup(
 export function removeGroup(
   requestOptions: IGroupIdRequestOptions
 ): Promise<any> {
-  const url = `${getPortalUrl(
-    requestOptions
-  )}/community/groups/${requestOptions.id}/delete`;
+  const url = `${getPortalUrl(requestOptions)}/community/groups/${
+    requestOptions.id
```
Comment (tomwayson) [inline packages/arcgis-rest-groups/src/groups.ts]: TIL - you can have line breaks w/in `${}` in template literals w/o adding line breaks to the output.

**Comment by tomwayson** [comment]
If we don't rename the existing interfaces, then this would not have to be a breaking change, right?

If not, I say we lump these in w/ either a 1.0.3 release, or a 1.1.0 release w/ #115

File: tslint.json [hunk 2]
```diff
@@ -8,6 +8,7 @@
     "strict-type-predicates": false,
     "ordered-imports": ["any"],
     "only-arrow-functions": [false],
-    "object-literal-sort-keys": false
+    "object-literal-sort-keys": false,
+    "interface-name": [false, "always-prefix"]
```
Comment (tomwayson) [inline tslint.json]: I think we want to remove this now that we're back to `I`ing everything, right?

Comment (jgravois) [reply tslint.json]: definitely. fixed.

**Comment by tomwayson** [comment]
This LGTM now, so merge and include in 1.0.3, or wait for 1.1.0?

**Comment by jgravois** [comment]
> merge and include in 1.0.3, or wait for 1.1.0?

I'm fine with either. i just want to make sure and include #120 in `v1.0.3`

**Comment by tomwayson** [comment]
If 1.0.3 is eminent, I'd say let this be the first PR of 1.1.0

**Pull request merged**